src/index.js

```diff
@@ -108,11 +124,7 @@ function wrapAdapter (PouchDB, HttpPouch) {
 
     db.originalFetch = db.fetch || PouchDB.fetch;
     db.fetch = async (url, opts = {}) => {
-      let session = getExistingSession(db);
-      if (!session) {
-        await authenticate(db);
-        session = getExistingSession(db);
-      }
+      const session = await validateSession(db, await getSession(db));
```

Comment:
I think this be easier to read if `getSession` called `validateSession` before it returned, as in, I'd never expect getSession to return an invalid session.

Reply:
I believe I also need to validate outside of the get, because I need to validate that the session has not expired for every request, which is separate from getting the session. I suppose I can make a function to wrap these two calls but validation and retrieval have to be separate :(

Reply:
Unless I'm missing something you're already calling `getSession` and `validateSession` every request, right?

So you could have getSession do something like...

```js

const isValid = (db, session) => {
  if (!session || !session.expires) {
    return false;
  }

  // TODO could just return this line, but maybe naming the var is more clear?
  const isExpired =  Date.now() > session.expires;
  if (isExpired) {
    return false;
  }

  return true;
};

const getValidSession = async (db) => {
  const sessionKey = getSessionKey(db);
  const session = sessions[sessionKey];

  if (isValid(session)) {
    return session;
  }

  invalidateSession(db); // this line is not necessary any more, because we're about to set the value to a promise
  sessions[sessionKey] = authenticate(db);
  return sessions[sessionKey];
};

```

I think this is functionally equivalent except you're not validating the fresh session (which I think is ok??). If you agree it's the same, then I think it's a little easier to read.

Reply:
I need to validate every time, because I'm evaluating whether a session is already expired. so validation needs to happen for every fetch. And not one time when we get the session, when we assume it's not expired.

Reply:
That code validates every time.

Reply:
Cool! Thanks for the suggestion. I agree it's probably not necessary to validate cookie expiry immediately after getting it.